proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2357 @@
+this section will show.
+
+Concretely, how do we ensure that the sockets from the example API are
+always closed, even in presence of exceptions? This boils down to how
+the ``RIO`` monad is implemented. Below is a sketch of one possible
+implementation of ``RIO`` (see `here
+<https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs>`_
+for a detailed implementation).
+
+First, note that since Haskell programs are of type ``IO ()``, we need a
+way to run ``RIO`` in an ``IO`` computation, this is provided by the
+function
+
+::
+
+  runRIO :: RIO (Unrestricted a) -> IO a
+
+In order to achieve resource safety in presence of exception, ``runRIO``
+is tasked with releasing any live resource in case of exception.
+
+To implement this, ``RIO`` keeps a table of release actions, to be used
+in case of exceptions. Each resource implemented in the ``RIO``
+abstraction registers a release action in the release action table
+when they are acquired.
+
+If no exception occurs, then all resources have been released by the
+program. In case an exception occurs, the program jumps to the handler
+installed by ``runRIO``, which releases the leftover resources.
+
+An alternative strategy would be to add terminators on every resources
+acquired in ``RIO``. Release in the non-exceptional case would still
+be performed by the program, and the GC would be responsible for
+releasing resources in case of exception. The release in case of
+exception would be, however, less timely.
+
+Can ``RIO`` have a ``catch``?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is possible to catch exceptions inside of ``RIO``. But in order to
+ensure resource safety, the type cannot be linear:
+
+::
+
+  catchL :: Exception e
+         => RIO (Unrestricted a)
+	 -> (e -> RIO (Unrestricted a))
+	 -> RIO (Unrestricted a)
+
+That is: no linear resource previously allocated (in particular linear
+variables which are not ``RIO`` resources) can be referenced in the
+body or the handler, and no resource allocated in the body or handler
+can be returned. In effect, ``catchL`` delimits a new scope, in which
+linear resources are isolated. To implement ``catchL``, we simply give
+it its own release action table, so that in case of exceptions all the
+local resources are released by ``catchL``, as ``runRIO`` does, before
+the handler is called. The original release action table is then
+reinstated.
+
+With this implementation, it is clear that capturing linear resources
+from the outside scope would compromise timely release, and returning
+locally acquired resources would leak resources if an exception
+occurs.
+
+The latter restriction can be lifted as follows: instead of
+reinstating the original release action table in the non-exceptional
+case, instate the *union* of the original table and the local one. In
+this case the type of ``catchL`` would be the following:
+
+::
+
+  catchL :: Exception e
+         => RIO a -> (e -> RIO a) -> RIO a
```

Comment:
So the problem I imagine with this is with translating code like

```
  flip catch (\e -> .. handle exceptions ..) $
      bracket allocateResource deallocateResource $ \r -> do
         ... do some operations with r ...
```

would be something like

```
runRIO $ do
  flip catchL (\e -> .. handle exceptions ..) $ do
    r <- allocateResource
    ... do some operations with r ...
    deallocateResource r
```

And now the resource will not be released until after the exception handler completes, which can be an arbitrarily long time. The only way to fix this would be to use nested `runRIO`s, which might not be possible because you can't have resources that span the nested scope.

Reply:
I don't think so: in this implementation, if there is an exception in the body, you would still deallocate every resource from the release table before entering the exception handler.

It's only in case of non-exceptional return from `catchL` that `r` could be returned, hence its exceptional deallocation be entrusted to the larger release table.

Does this make sense?

Reply:
Hmm, I think I was confused. I might still be confused. Could you say exactly which resources are released under what circumstances? 

If a linear resource is allocated before the `catchL` can it be released in both the body and the handler of the `catchL`? 

Are there any reasons you might prefer the first `catchL` over this `catchL`?